### ISSUE TITLE
[Networking] Bittorrent over libp2p (in rust)

### ISSUE THREAD

**Issue by gnunicorn** [substrate-developer-hub/hacktoberfest#16]
Implement the [bittorrent data exchange protocol](http://www.bittorrent.org/beps/bep_0003.html) as a rust-libp2p-protocol.

For mentoring this issue please contact: [me @gnunicorn](https://substrate.dev/hacktoberfest/hackers/gnunicorn.html)

## Proof of done
These following tasks must have been completed in order for this to be considered done:
 - [ ] Create an external crate for the libp2p-bittorrent-protocol implemented on top of rust-libp2
  - [ ] be sure to include an `example` on how to use that protocol
 - [ ] Raise a PR against this Repository, adding your profile info under `participants`
 - [ ] Raise a PR containing `fixes #16` with your submission under `/submissions` against this repository with a git submodule to your open github repo.

**Comment by mattsse** [comment]
I started working on this during the weekend, but mostly research since I didn't work with lip2p yet.
If I understood correctly a rust-libp2p-protocol implementation should use a protbuf based message scheme instead of bencoding?
So far I added the message types and protbuf conversions: [https://github.com/mattsse/libp2p-bittorrent-protocol](https://github.com/mattsse/libp2p-bittorrent-protocol)

I also added tracker definitions which probably should have its own protocol and converts tracker requests to DHT lookups using kademlia.

**Comment by gnunicorn** [comment]
Hey, @mattsse,
welcome to hacktoberfest, happy to have you aboard :tada: !

> If I understood correctly a rust-libp2p-protocol implementation should use a protbuf based message scheme instead of bencoding?

Why though? I was thinking of more or less directly implement bittorrent on libp2p. Why not use bencoding if that is what bittorrent is using? Protobuf is mostly being dropped for most libp2p protocols (because it is way more complicated than needed)..

In substrate we mostly use [Scale encoding](https://substrate.dev/docs/en/overview/low-level-data-format) and it has some [pretty handy rust implementation](https://github.com/paritytech/parity-scale-codec). So if that makes it easier, you can also start out encoding the messages  that way.

But in the end, if two nodes can talk the protocol itself with one another and exchange data, I do not really care...

Double plus plus, though, if the encoding would be configurable ... maybe as a feature-flag?

**Comment by mattsse** [comment]
Thanks for your quick reply. Happy to be aboard!

> I was thinking of more or less directly implement bittorrent on libp2p.

Got it. I guess I was confused by the explicit reference to protobuf in the docs. I'll drop it in support of bencoding with the support to configure encoding via feature flags in mind.

Even if [serde-bencode](https://github.com/toby/serde-bencode) is stale I'll probably start working with it or a fork for the time being.

**Comment by gnunicorn** [comment]
@mattsse there is also bendy, which seems to be more actively developed: https://crates.io/crates/bendy

**Comment by mattsse** [comment]
Hi @gnunicorn ,
unfortunately I'am not quite down with the assignment and can't finish it in time for hacktoberfest.
Progress so far is

- [x]  Reading writing torrent files
- [x] Implementation of the bittorrent Peer Wire Protocol Messages and [tokio codec](https://github.com/mattsse/libp2p-bittorrent-protocol/blob/master/src/proto/codec.rs) which means, other codecs could be supported by simply chaning the codec used in the [In- and `OutboundUpgrade`s](https://github.com/mattsse/libp2p-bittorrent-protocol/blob/master/src/proto/mod.rs)
- [ ] I guess I have the overall design (`handler, behavior, diskmanager, torrentpool`) pretty much down so far but I still have a lot of loose ends that I need to tie up.
- [ ] examples

Should I submit a WIP PR now to let it count for hacktoberfest, if this makes any difference at all?
Or wait until it's done (mid of next week probably)

**Comment by gnunicorn** [comment]
No worries, @mattsse, I got your contribution recorded, just submit it next week, when it is ready.

But make sure to add your contact details through the signup form by today latest if you haven't yet!

**Comment by mattsse** [comment]
Hi @gnunicorn 
this took me alot longer than anticipated at first, but I have a working [example ](https://github.com/mattsse/libp2p-bittorrent-protocol/blob/master/examples/bittorrent.rs) now.
There are still some rough edges that need cleanup and the Disk IO handling is somewhat hackish.

I'd also would like to add [trackerless torrents via DHT](https://www.bittorrent.org/beps/bep_0005.html), who could this be achieved using Kademlia? with an `upgrade::EitherUpgrade<BitTorrentProtocolConfig,KademliaProtocolConfig>` in the Handler?

**Comment by gnunicorn** [comment]
@mattsse Yes, Kademlia is a DHT system. I am not exactly sure how this works myself, maybe @tomaka knows.

Regarding your project: This is very awesome, thank you very much. Your swag will be on its way shortly. What are your plans with it now? What is needed before publishing it on crates.io? Or would you rather want to merge it into libp2p-rust?

Again, great work :tada:!

**Comment by tomaka** [comment]
The idea is not to group upgrades or protocol handlers together, but rather to group `NetworkBehaviour`s together.

There is an example here for how to do that, using a custom derive: https://github.com/paritytech/substrate/blob/f44873dce009b1b651434c836b5be09d546f1915/client/network/src/behaviour.rs#L34-L49
